### PR TITLE
Ticket 57: Redacted wildcards should not be allowed

### PR DESCRIPTION
**Pull request by eranmes**
Some modifications to the redaction text as well as indication that redacted wildcards are not allowed.
Commits are separate to ease review, will squash when (if!) they are accepted.

### CODE REVIEW

**Comment by eranmes** [comment]
I was using the same term ("to secure") used in a preceding a paragraph.
Replaced with your suggested text.

**Comment by robstradling** [comment]
Sorry for the branch conflicts on this PR, Eran.  These are due to https://github.com/google/certificate-transparency-rfcs/commit/d71d7707a3eeb5707cf5048c3849b068e477038c

**Comment by eranmes** [comment]
@robstradling PTAL, addressed your comments.

File: rfc6962-bis.xml [hunk 1]
```diff
@@ -356,7 +356,7 @@ d0 d1   d2 d3           d0 d1   d2 d3  d4 d5</artwork>
...
+          When a precertificate contains one or more redacted labels, a non-critical extension (OID 1.3.6.1.4.1.11129.2.4.6, whose extnValue OCTET STRING contains an ASN.1 SEQUENCE OF INTEGERs) MUST be added to the corresponding certificate: the first INTEGER indicates the total number of ? labels in the precertificate's first DNS-ID; the second INTEGER does the same for the precertificate's second DNS-ID; etc. There MUST NOT be more INTEGERs than there are DNS-IDs. If there are fewer INTEGERs than there are DNS-IDs, the shortfall is made up by implicitly repeating the last INTEGER. Each INTEGER MUST have a value of zero or more. <xref target="reconstructing_tbscertificate"/> describes how TLS clients should reconstruct the precertificate from a final certificate.
```
Comment (robstradling) [inline rfc6962-bis.xml]: Nit: Please enclose the ? in a <spanx>

Comment (robstradling) [inline rfc6962-bis.xml]: It's not "the precertificate" that is reconstructed by TLS clients.  It's "the TBSCertificate component of the precertificate", as the current text says.
Also, now that precertificates are not X.509 certificate, I'm not keen on "final certificate".  Just "certificate" will do.

Comment (robstradling) [inline rfc6962-bis.xml]: I suggest keeping the "The purpose of this extension..." sentence.  You could simply change...
'accurately reconstruct the TBSCertificate component of the precertificate'
to
'<xref target="reconstructing_tbscertificate">accurately reconstruct the TBSCertificate component of the precertificate</xref>'

Comment (robstradling) [inline rfc6962-bis.xml]: (Ignore the backslashes - just trying to persuade GitHub to display what I originally typed :-) )

File: rfc6962-bis.xml [hunk 2]
```diff
@@ -1314,6 +1314,20 @@ but it is expected there will be a variety.
             Since each received SCT may be valid either for the server certificate itself or for a <xref target="name_constrained">suitable name-constrained intermediate</xref>, TLS clients MAY need to use trial and error to determine which certificate in the chain is the certificate to which a received SCT corresponds.
           </t>
         </section>
+        <section title="Reconstructing the TBSCertificate" anchor="reconstructing_tbscertificate">
```
Comment (robstradling) [inline rfc6962-bis.xml]: The TBSCertificate is needed for validating both SCTs and inclusion proofs, so it seems a little odd to put this section inbetween the "Validating SCTs" and "Validating inclusion proofs" sections.
I suggest moving this section before the "Validating SCTs" section.

Comment (robstradling) [inline rfc6962-bis.xml]: Also - and perhaps this should be done in a different PR - I think the "To validate an SCT, a TLS client computes the signature input from the SCT data..." sentence in the "Validating SCTs" section could perhaps do with some clarification.  The reconstructed TBSCertificate should be mentioned here, I think.

File: rfc6962-bis.xml [hunk 3]
```diff
@@ -1314,6 +1314,20 @@ but it is expected there will be a variety.
...
+          <t>
+            To reconstruct the TBSCertificate component of a precertificate from a certificate, TLS clients should:
+          <list style="symbols">
```
Comment (robstradling) [inline rfc6962-bis.xml]: Indentation?

File: rfc6962-bis.xml [hunk 4]
```diff
@@ -1327,7 +1341,9 @@ but it is expected there will be a variety.
         </section>
         <section title="Evaluating compliance">
           <t>
-            To be considered compliant, a certificate MUST be accompanied by at least one valid SCT or at least one valid inclusion proof. A certificate not accompanied by any valid SCTs or any valid inclusion proofs MUST NOT be considered compliant by TLS clients. However, specifying the TLS clients' behavior once compliance or non-compliance has been determined (for example, whether a certificate should be rejected due to non-compliance) is outside the scope of this document.
+            To be considered compliant, a certificate MUST be accompanied by at least one valid SCT or at least one valid inclusion proof. A certificate not accompanied by any valid SCTs or any valid inclusion proofs MUST NOT be considered compliant by TLS clients.
+          </t>
```
Comment (robstradling) [inline rfc6962-bis.xml]: Not sure why this needs splitting into two paragraphs.

File: rfc6962-bis.xml [hunk 5]
```diff
@@ -1327,7 +1341,9 @@ but it is expected there will be a variety.
...
+          <t>However, specifying the TLS clients' behavior once compliance or non-compliance has been determined (for example, whether a certificate should be rejected due to non-compliance) is outside the scope of this document.
```
Comment (robstradling) [inline rfc6962-bis.xml]: Actually, I suggest removing this sentence completely, because it seems inconsistent with the recently added "Handling of Non-compliance" section.  @benlaurie Does that work for you?

File: rfc6962-bis.xml (hunk 1 above)
Comment (eranmes) [inline rfc6962-bis.xml]: I ended up leaving most of the sentence and appending the reference to the end - would work better in the text version of the RFC IMHO.

File: rfc6962-bis.xml (hunk 3 above)
Comment (eranmes) [inline rfc6962-bis.xml]: Done.

File: rfc6962-bis.xml (hunk 2 above)
Comment (eranmes) [inline rfc6962-bis.xml]: Moved the section - not entirely sure how you want to mention the reconstructed TBSCertificate in the  Validating SCTs section.

File: rfc6962-bis.xml (hunk 5 above)
Comment (eranmes) [inline rfc6962-bis.xml]: Removed as suggested.

**Comment by eranmes** [comment]
@robstradling PTAL, addressed all your comments.

**Comment by robstradling** [comment]
Indentation in the "Reconstructing the TBSCertificate" section still needs some tweaks.

**Comment by robstradling** [comment]
The "Merkle Tree Leaves" section says:
`(Note that a precertificate's TBSCertificate can be reconstructed from the issued certificate's TBSCertificate by redacting the domain name labels indicated by the redacted labels extension, and deleting the SCT list extension and redacted labels extension).`
Please would you change this to:
`(Note that a precertificate's TBSCertificate can be reconstructed from the corresponding certificate as described in <xref target="reconstructing_tbscertificate"/>.`

File: rfc6962-bis.xml (hunk 2 above)
Comment (robstradling) [inline rfc6962-bis.xml]: I'm not sure either.  :-)
It's just that "computes the signature input from the SCT data" seems like a high-level description that could usefully be fleshed out in the document to help implementers.

**Comment by eranmes** [comment]
Done both, PTAL.

**Comment by robstradling** [comment]
Sorry for the conflicts.  They're due to https://github.com/google/certificate-transparency-rfcs/pull/128

**Comment by eranmes** [comment]
Squashed, pushing now.
